File: coco/src/main/java/coco/ep/m7i2/Power.java
package coco.ep.m7i2;

import coco.ep.Exp;
import util.Node;
import util.Tree;

import java.util.List;

public interface Power<FT> extends coco.ep.i2.Power<FT>, coco.ep.m7i2.Exp<FT>, Factory<FT> {
	void setLeft(coco.ep.Exp<FT> left);
	void setRight(coco.ep.Exp<FT> right);

    default coco.ep.Exp<FT> multby(coco.ep.Exp<FT> other) {return this.mult(this, convert(other)); }
    default coco.ep.Exp<FT> powby(coco.ep.Exp<FT> other) {
        return this.power(this, convert(other));
    }

    default void truncate (int level) {
		if (level > 1) {
			convert(getLeft()).truncate(level-1);
			convert(getRight()).truncate(level-1);
		} else {
			setLeft(this.lit(convert(getLeft()).eval()));
			setRight(this.lit(convert(getRight()).eval()));
		}
	}
	
    default Exp<FT> simplify() {
        if (convert(getRight()).eval().equals(0.0)) {
            return this.lit(1.0);
        } else if (convert(getLeft()).eval().equals(1.0)) {
            return this.lit(1.0);
        } else if (convert(getRight()).eval().equals(1.0)) {
            return convert(getLeft()).simplify();
        } else {
            return this.power(convert(getLeft()).simplify(), convert(getRight()).simplify());
        }
    }

    default List<Double> collect() {
        return java.util.stream.Stream.concat(convert(getLeft()).collect().stream(), convert(getRight()).collect().stream()).collect(java.util.stream.Collectors.toList());
    }

    default Tree astree() {
        return new Node(convert(this).id(), convert(getLeft()).astree(), convert(getRight()).astree());
    }

    default Integer id() {
        return 77306085;
    }

    default Boolean equals(Exp<FT> other) {
        return convert(this).astree().equals(convert(other).astree());
    }

    default Boolean isPower(coco.ep.Exp<FT> left, coco.ep.Exp<FT> right) {
        return convert(left).eql(getLeft()) && convert(right).eql(getRight());
    }

    default Boolean eql(coco.ep.Exp<FT> that) {
	    return convert(that).isPower(convert(getLeft()), convert(getRight()));
    }
}

File: coco_csharp/CoCo/EP/M7I2/Power.cs
using System.Collections.Generic;

namespace EP.M7I2 {
    public interface Power<FT> : Exp<FT>, EP.I2.Power<FT> {
        void setLeft(EP.Exp<FT> left);
        void setRight(EP.Exp<FT> right);

         EP.Exp<FT> EP.I1.Exp<FT>.multby(EP.Exp<FT> other) {return mult(this, convert(other)); }
         EP.Exp<FT> EP.M7.Exp<FT>.powby(EP.Exp<FT> other) {
            return power(this, convert(other));
        }

         void EP.M4.Exp<FT>.truncate (int level) {
            if (level > 1) {
                convert(getLeft()).truncate(level-1);
                convert(getRight()).truncate(level-1);
            } else {
                setLeft(lit(convert(getLeft()).eval()));
                setRight(lit(convert(getRight()).eval()));
            }
        }
        
        EP.Exp<FT> EP.M4.Exp<FT>.simplify() {
            if (convert(getRight()).eval() == (0.0)) {
                return lit(1.0);
            } else if (convert(getLeft()).eval() == (1.0)) {
                return lit(1.0);
            } else if (convert(getRight()).eval() == (1.0)) {
                return convert(getLeft()).simplify();
            } else {
                return power(convert(getLeft()).simplify(), convert(getRight()).simplify());
            }
        }

        List<double> EP.M4.Exp<FT>.collect() {
            List<double> result = new List<double>();
            result.AddRange(convert(getLeft()).collect());
            result.AddRange(convert(getRight()).collect());
            return result;
        }

        EP.M5.Tree EP.M5.Exp<FT>.astree() {
            return new M5.Node(convert(this).id(), convert(getLeft()).astree(), convert(getRight()).astree());
        }

        int EP.M5.Exp<FT>.id() {
            return 77306085;
        }

        bool EP.M6.Exp<FT>.equals(EP.Exp<FT> other) {
            return convert(this).astree().Equals(convert(other).astree());
        }

        bool EP.M7I2.Exp<FT>.isPower(EP.Exp<FT> left, EP.Exp<FT> right) {
            return convert(left).eql(getLeft()) && convert(right).eql(getRight());
        }

        bool EP.M6.Exp<FT>.eql(EP.Exp<FT> that) {
            return convert(that).isPower(convert(getLeft()), convert(getRight()));
        }
    }
}